### ISSUE TITLE
App crashing on android 12 (api32)

### ISSUE THREAD

**Issue by glpnk** [wh201906/mEDIFIER#1]
App is crash on bt rfcomm and ble discovery

UPD: both arm64 and universal

**Comment by glpnk** [comment]
Also how you discover available modes for your headphones? I`m just streamed btsnoop log to wireshark over adb and triggered all available and unavailable (via some activity launcher) functions from app, and some bruteforce with raw commands for other functions

**Comment by wh201906** [comment]
> Also how you discover available modes for your headphones?

I don't have a rooted phone so I cannot stream the btsnoop log to  wireshark. However I did get some information from the btsnoop log by `adb bugreport`.

In most of the cases I just use Logcat. The Edifier Connect logs all of the communications. 
Here is the log for sending data
```
2023-06-10 01:27:43.029 CommandData             com.edifier.edifierconnect           I  addCRC_sourceData:AA01D8
2023-06-10 01:27:43.029 CommandData             com.edifier.edifierconnect           I  addCRC_encryptData:AA01D8
2023-06-10 01:27:43.029 CommandData             com.edifier.edifierconnect           I  CRC_value:AA01D8219C,length:8604,data:AA01D8
2023-06-10 01:27:43.029 BaseFragment            com.edifier.edifierconnect           I  sendData2Serivce_length:5
2023-06-10 01:27:43.029 SendMessageManageProxy  com.edifier.edifierconnect           I  insertCommandIntoQueue_commandType:216,commandArray_length:5,isCalling:false
2023-06-10 01:27:43.031 SendMessageManageProxy  com.edifier.edifierconnect           I  insert_queue_command:AA01D8219C,queue_length:1
2023-06-10 01:27:43.031 SendMessageManageProxy  com.edifier.edifierconnect           I  sendDataToBleDevice 发送:AA01D8219C,isReSend:false,isFirst:true
2023-06-10 01:27:43.031 AppWidgetService        com.edifier.edifierconnect           I  ----sendData--- isConnectedUsb:false,isSppConnect:,isCalling:false
2023-06-10 01:27:43.032 AppWidgetService        com.edifier.edifierconnect           I  发送数据的16进制-->AA01D8219C
```
The last line is "Data to send in hex-->AA01D8219C"

Here is the log for receiving data
```
2023-06-10 01:27:43.316 AppWidgetService 接收原始数据 com.edifier.edifierconnect           I  : BB04C601000221A1
```

**Comment by wh201906** [comment]
> App is crash on bt rfcomm and ble discovery

Do you grant the location permission to the app? I uses an older Android API so the location permission is necessary.
https://developer.android.com/guide/topics/connectivity/bluetooth/permissions?hl=en#declare-android11-or-lower

Plus, I haven't implement the BLE connection yet.

**Comment by glpnk** [comment]
> Do you grant the location permission to the app?

Yeah, in both builds, checked all permissions (single time, foreground only, background), all crashed

Also in log `Pending exception java.lang.SecurityException: Need android.permission.BLUETOOTH_CONNECT permission for android.content.AttributionSource@3cb880b: AdapterService getBondedDevices`

Google broke backward compatibility and need explicit `BLUETOOTH_CONNECT` permission with API version check in manifest

**Comment by wh201906** [comment]
I see. Thanks!

**Comment by glpnk** [comment]
Do you plan to add settings in notification like in original app?

**Comment by wh201906** [comment]
I haven't planned it yet. I guess it's not easy to implement it in Qt.

**Comment by glpnk** [comment]
In flutter it look complicated too. I think it`s need to be raw java/kotlin code

**Comment by wh201906** [comment]
Indeed

**Comment by wh201906** [comment]
@glpnk Hi. Could you please try the latest release v0.0.3? I think the permission bug should be resolved.

**Comment by glpnk** [comment]
Solved!

Recommendation: add confirmation to "dangerous" buttons like power off and other

**Comment by wh201906** [comment]
@glpnk Would you mind telling me how you triggers all functions in the app? Now I want to create different interfaces for different devices.

**Comment by glpnk** [comment]
@wh201906 All documented commands [here](https://github.com/GlitchPunkWTF/edifier-proto#w820nb-commands).

Also you can create fancy looking documentation like [this](https://github.com/GlitchPunkWTF/edifier-proto/blob/master/models/w820nb.md) and open PR if you had another model of headphones and copy it to your repo. Will be great to collect all known information about Edifier commands in some convenient format.

About triggering functions: after connections I`m just sending [all get commands](https://github.com/GlitchPunkWTF/edifier-proto/blob/23fce8af4b380a968807893ae3f38ec7be5c7336/lib/app/edifier.dart#L11) and [parsing answers](https://github.com/GlitchPunkWTF/edifier-proto/blob/23fce8af4b380a968807893ae3f38ec7be5c7336/lib/app/edifier.dart#L51). If your packet exceed 20 bytes len you need to split it into [two parts](https://github.com/GlitchPunkWTF/edifier-proto/blob/23fce8af4b380a968807893ae3f38ec7be5c7336/lib/app/bluetooth.dart#L96) and merge two parts [on receive](https://github.com/GlitchPunkWTF/edifier-proto/blob/23fce8af4b380a968807893ae3f38ec7be5c7336/lib/app/bluetooth.dart#L77).

I don't know how to determine model but I guess it may be related with `d8` command

**Comment by wh201906** [comment]
Thanks

**Comment by glpnk** [comment]
I created this app because I got tired of using [python script](https://github.com/GlitchPunkWTF/edifier-proto/blob/master/poc.py) or `nRF Connect` app every time this headphones loosing settings after wrong power on and charger plugged when device is on. From readme I guess you have `W820NB Plus` model. Do you have same problems? It occurs because buttons get stuck and second attempt to power on leads to wiped settings

What do you think is encoded in `d8`?

Also how to compile app on linux?

**Comment by wh201906** [comment]
> Do you have same problems?

Yeah I got the exact same problem on W820NB Plus. But I'm not sure if it's caused by the buttons

> What do you think is encoded in `d8`?

I haven't look at it yet, but I got something interesting from the BLE servce UUID, which can be used to detect the model.

> Also how to compile app on linux?

You want to compile this app on Linux? or your flutter one?

**Comment by wh201906** [comment]
Would you like to start a private conversation? you can email me at wh201906@yandex.com

**Comment by glpnk** [comment]
Yeah but my default email service could be blocked by Yandex because of Roskomnadzor - Russian great firewall creator

**Comment by wh201906** [comment]
How about wh201906@yeah.net

**Comment by wh201906** [comment]
> I created this app because I got tired of using [python script](https://github.com/GlitchPunkWTF/edifier-proto/blob/master/poc.py) or `nRF Connect` app every time this headphones loosing settings after wrong power on and charger plugged when device is on.

I've add a feature to save/restore all settings from a json file. Hope it helps.
